src/sentry/release_health/metrics_sessions_v2.py

```diff
@@ -312,15 +311,72 @@ def get_values(
             ]
 
 
+class _LimitState:
+    """Keeps track of how to limit each query
+    Before the first query, limiting_conditions = None, so the first query
+    will apply an ORDER BY ... LIMIT.
+    After that, the groups returned by the first query will be added as a
+    WHERE clause to any secondary query.
+    """
+
+    def __init__(self) -> None:
+        self._initialized = False
+        self.skip_columns: List[Column] = []  # Will not be added to limiting conditions
+
+    def update(self, groupby: Collection[Column], snuba_rows: _SnubaData) -> None:
+        if self._initialized:
+            return
+
+        # Only "totals" queries may set the limiting conditions:
+        assert Column(TS_COL_GROUP) not in groupby
+
+        def sort_key(col: Column) -> str:
+            return col.name  # type: ignore. But why?
+
+        groupby = sorted(groupby, key=sort_key)
+
+        self._groupby = sorted(groupby, key=sort_key)
+        self._groups = [{column.name: row[column.name] for column in groupby} for row in snuba_rows]
+
+        self._initialized = True
+
+    @property
+    def limiting_conditions(self) -> Optional[List[Condition]]:
+        if not self._initialized:
+            # First query may run without limiting conditions:
+            return None
+
+        group_columns = [col for col in self._groupby if col not in self.skip_columns]
+
+        if not group_columns:
+            return []
+
+        # Create conditions from the groups in group by
+        group_values = [
+            Function("tuple", [row[column.name] for column in group_columns])
+            for row in self._groups
+        ]
+        return [
+            # E.g. (release, environment) IN [(1, 2), (3, 4), ...]
+            Condition(Function("tuple", group_columns), Op.IN, group_values)
+        ] + [
+            # These conditions are redundant but might lead to better query performance
+            # Eg. [release IN [1, 3]], [environment IN [2, 4]]
+            Condition(column, Op.IN, [row[column.name] for row in self._groups])
+            for column in group_columns
+        ]
+
+
 def _get_snuba_query(
     org_id: int,
     query: QueryDefinition,
     entity_key: EntityKey,
     metric_id: int,
-    columns: Sequence[SelectableExpression],
+    columns: List[SelectableExpression],
     series: bool,
+    limit_state: _LimitState,
```

Comment:
I am not a big fan of passing `limit_state` here because you basically only need it here for the `limiting_conditions`, and you don't mutate the object whatsoever
So passing this in is pretty much redundant to passing `extra_conditions` in the next arg.. I would suggest unifying this somehow

Reply:
Now passing `limiting_conditions` instead. Could not unite with `extra_condition` though, because some logic depends on `limiting_conditions` being none.

Reply:
I see.. In this case, I might have give you bad advice sorry.. I think you should pass the `LimitState` and rather than checking if `limiting_conditions` is None, check for the `initialized` flag